### ISSUE TITLE
Different results between individual and batch processing in Chronos-Bolt model

### ISSUE THREAD

**Issue by StatMixedML** [amazon-science/chronos-forecasting#262]
## Description
When using Chronos Base model, there are inconsistent forecasting results between processing time series individually versus passing them as a list. This behavior is not observed as severely for Chronos Bolt, which seems to maintain consistency across both context types. The issue seems to be most pronounced for a low number of samples drawn from the distribution and occurs irrespective of the context-length for each series (at first I was assuming the issue is due to the NaN-padding when creating batches for series of different length, but it also occurs for series with same context-length). The issue gets more pronounced the more series we pass to the model via the context.

## Current Behavior
- Different forecasting results between individual and batch processing for Chronos Base
- Issue persists even when series have identical lengths
- Appears using the "base" size, but haven't tested it with "tiny" or "large"

## Expected Behavior
Consistent forecasting results regardless of context type (individual vs batch).

## Additional Context
- Tested with series of both different and identical lengths
- Behavior is consistent across multiple runs with fixed random seeds

## Steps to Reproduce
1. Generate forecasts using two methods:
   a. Process each series individually 
   b. Process all series as a batch list
2. Compare the forecasting results

Minimal reproduction code attached.
[chronos_gitissue.zip](https://github.com/user-attachments/files/18421184/chronos_gitissue.zip)

**Environment description**
Operating system: Windows 10
Python version: 3.11
CUDA version: 11.8
PyTorch version: 2.1.1
HuggingFace Hub version: 0.22.1
Chronos Version: 1.2.0
Chronos-Forecasting Version: 1.4.1

### Chronos Base
![image](https://github.com/user-attachments/assets/601964f2-ccfa-42c8-b4ee-8b4135a76496)

### Chrons Bolt
![image](https://github.com/user-attachments/assets/ad24c5b5-d8ea-43b9-9d82-0ade7f0213dd)

**Comment by StatMixedML** [comment]
As a rejoinder: The issue is even more pronounced using the [VN1-Data](https://github.com/Nixtla/nixtlar/tree/master/experiments/vn1-forecasting-competition). When using this more intermittent dataset, both Chronos Base and Chronos Bolt yield diverging results based on whether the series are processed individually or passed as a list.

## Plots (VN1 Data)

### Chronos Base
![image](https://github.com/user-attachments/assets/e0a742de-9474-444c-9dcc-8e96f6a9ea2b)

### Chronos Bolt
![image](https://github.com/user-attachments/assets/5b995954-1a8e-4859-bc69-068dac5fdbff)

**Comment by lostella** [comment]
@StatMixedML the images for the VN1 dataset appear identical, and both mention the Bolt model. Are you sure this is correct?

**Comment by lostella** [comment]
@StatMixedML for the first example (tourism data) have you tried increasing the number of samples to see if the difference reduces for `chronos-t5-base`? I see you're using 10, it would be nice to try 20 and 50 to see if that reduces the effect (maybe even on a single series, since the more the samples, the slower the prediction). If the difference reduces as expected, than this may be an indication that batch vs loop results in different random sequence being drawn under the hood, no matter the seed initialization

**Comment by StatMixedML** [comment]
> @StatMixedML the images for the VN1 dataset appear identical, and both mention the Bolt model. Are you sure this is correct?

@lostella Thanks for pointing me to this. I replaced the redundant image with the correct one.

**Comment by StatMixedML** [comment]
> @StatMixedML for the first example (tourism data) have you tried increasing the number of samples to see if the difference reduces for `chronos-t5-base`? I see you're using 10, it would be nice to try 20 and 50 to see if that reduces the effect (maybe even on a single series, since the more the samples, the slower the prediction). If the difference reduces as expected, than this may be an indication that batch vs loop results in different random sequence being drawn under the hood, no matter the seed initialization

I am currently running the evaluation also with a higher number of samples to see if the effect is still as pronounced. Yet, since Chronos Bolt directly forecasts quantiles, not sure why it also happens for Bolt.

**Comment by lostella** [comment]
> Yet, since Chronos Bolt directly forecasts quantiles, not sure why it also happens for Bolt? Also, why would the differences for Bolt only appear for the VN1 dataset?

Yes, this would indeed be very strange, I'll try to reproduce it.

Otherwise, I believe the behaviour you see with `chronos-t5-base` is to be expected. Think about what happens under the hood in the two cases, if you're drawing 10 samples, prediction length 2, for multiple series:
- When looping, the first 20 random numbers drawn will be "spent" on the first series (10 samples for the first step, 10 samples for the second step).
- When processing in batch, the first 20 random numbers drawn will go into the first step of the first two series.
That's why I believe this is expected behaviour, and the only way to make the two cases more similar is to draw more samples.

**Comment by StatMixedML** [comment]
> @StatMixedML for the first example (tourism data) have you tried increasing the number of samples to see if the difference reduces for `chronos-t5-base`? I see you're using 10, it would be nice to try 20 and 50 to see if that reduces the effect (maybe even on a single series, since the more the samples, the slower the prediction). If the difference reduces as expected, than this may be an indication that batch vs loop results in different random sequence being drawn under the hood, no matter the seed initialization

So the number of samples indeed reduces the effect of the discrepancies, see the screenshots below. Not sure though if that is reducing my concerns. Why should the number of samples and the way you feed the context to the model have an influence on the actual forecast? Sur, the higher the samples the better the distribution is approximated and the more stable the corresponding mean and quantiles etc. However, sampling 10 observations and feeding a single series should give the same results when sampling 10 observations and feeding 100 series.

![image](https://github.com/user-attachments/assets/0c215cc0-a827-4008-bbdf-30ba346e0a15)

![image](https://github.com/user-attachments/assets/c2ebb4c2-cfb1-4e4e-bd5b-4cace5006946)

**Comment by lostella** [comment]
> Why should the number of samples and the way you feed the context to the model have an influence on the actual forecast?

See my previous message https://github.com/amazon-science/chronos-forecasting/issues/262#issuecomment-2592596308

**Comment by StatMixedML** [comment]
> > Why should the number of samples and the way you feed the context to the model have an influence on the actual forecast?
> 
> See my previous message [#262 (comment)](https://github.com/amazon-science/chronos-forecasting/issues/262#issuecomment-2592596308)

But why not use a context-type independent sampling, similar as used in DeepAR or any other probabilistic forecasting model? Or would the same issue also apply for GluonTS models?

**Comment by lostella** [comment]
Because we rely on the sampling procedure of the T5 implementation in the HuggingFace transformers package (see [here](https://github.com/amazon-science/chronos-forecasting/blob/ad410c9c0ae0d499aeec9a7af09b0636844b6274/src/chronos/chronos.py#L347-L361)), and I believe we do not have control over how the random samples being drawn are used in terms of batch vs time dimension (but I might be wrong, it would be cool if we could change it and keep the performance benefits of doing inference in batch).

**Comment by lostella** [comment]
Specifically, this is where the `GenerationMixin` class in HF transformers draws the tokens for the next step using `torch.multinomial`: see [code](https://github.com/huggingface/transformers/blob/12dfd99007dd803b3dfab5fe2ec66260b8007a23/src/transformers/generation/utils.py#L3298-L3300). For language it makes sense, because you probably want to terminate different sentences at different points. In our case, we always need to draw `batch_size x prediction_length` random numbers (in the simple case where `num_samples == 1`), so we could draw those, transpose the matrix, and get the same random sequence as we would get when looping through the series one by one for prediction. But then again, I don't think we have that kind of control.

**Comment by StatMixedML** [comment]
> Specifically, this is where the `GenerationMixin` class in HF transformers draws the tokens for the next step using `torch.multinomial`: see [code](https://github.com/huggingface/transformers/blob/12dfd99007dd803b3dfab5fe2ec66260b8007a23/src/transformers/generation/utils.py#L3298-L3300). For language it makes sense, because you probably want to terminate different sentences at different points. In our case, we always need to draw `batch_size x prediction_length` random numbers (say in the simple case where `num_samples == 1`), so we could draw those, transpose the matrix, and get the same random sequence as we would get when looping through the series one by one for prediction. But then again, I don't think we have that kind of control.

Thanks for pointing me to this! Will have a closer look.

Have you had the time to look into the VN1 Chronos-Bolt issue? Do you think that the scale of the time series has an effect, since this would be one obvious reason for the differences?

**Comment by lostella** [comment]
@StatMixedML I cannot reproduce the issue with Chronos-Bolt. The following minimal snippet can be adapted to other datasets, but the bottom line I see from electricity is that: there is a difference between the two modes, but the relative error appears to be large for very small numbers.

```python
import torch
from chronos import BaseChronosPipeline
from gluonts.dataset.repository import get_dataset

model = BaseChronosPipeline.from_pretrained("amazon/chronos-bolt-base")

dataset = list(get_dataset("electricity").train)

subset = [
    torch.tensor(dataset[0]["target"][-1500:]),
    torch.tensor(dataset[1]["target"][-600:])
]

# comment/uncomment the following to run on the whole dataset
subset = [torch.tensor(entry["target"]) for entry in dataset]

_, pred_batch = model.predict_quantiles(subset)

temp = []
for series in subset:
    _, new = model.predict_quantiles(series)
    temp.append(new)
pred_loop = torch.concat(temp, dim=0)

error = torch.abs(pred_batch - pred_loop)
print(f"     max absolute error: {torch.max(error)}")
print(f"     max relative error: {torch.max(error / torch.abs(pred_loop))}")
print(f"max mixed-abs-rel error: {torch.max(error / (1 + torch.abs(pred_loop)))}")
```

outputs:

```
     max absolute error: 0.1015625
     max relative error: 0.15625
max mixed-abs-rel error: 8.742192585486919e-05
```

**Comment by lostella** [comment]
It could be that on intermittent data, the scaling mechanism of Chronos-Bolt plays a different role and weird things happen. If there's an easy way to load the VN1 dataset in the script I have above, this is easy to check.

**Comment by StatMixedML** [comment]
Please find the VN1 example attached, including notebook and data. [Chronos_VN1.zip](https://github.com/user-attachments/files/18437274/Chronos_VN1.zip)

 Results are attached and as above.

![image](https://github.com/user-attachments/assets/76372734-3d92-4435-ae70-f8dd0e0776a7)

**Comment by lostella** [comment]
@StatMixedML if I print the first ten rows of the forecast dataframes with

```python
print(chronos_fcst_ind.head(10).to_markdown())
print(chronos_fcst_batch.head(10).to_markdown())
```

I get suspiciously close (if not identical) values in the `fcst` column, but differences in the `unique_id` and `y` columns. These seem to be used by the plotting logic. Are you sure there are no bugs in the pandas logic somewhere?

```
|    | ds                  | unique_id   |    fcst |   y | model                   |
|---:|:--------------------|:------------|--------:|----:|:------------------------|
|  0 | 2024-01-08 00:00:00 | 0/1/10705   | 2.85593 |   0 | Chronos-Bolt Individual |
|  1 | 2024-01-15 00:00:00 | 0/1/10705   | 2.75801 |   1 | Chronos-Bolt Individual |
|  2 | 2024-01-22 00:00:00 | 0/1/10705   | 2.65326 |   0 | Chronos-Bolt Individual |
|  3 | 2024-01-29 00:00:00 | 0/1/10705   | 2.53008 |   1 | Chronos-Bolt Individual |
|  4 | 2024-02-05 00:00:00 | 0/1/10705   | 2.43385 |   2 | Chronos-Bolt Individual |
|  5 | 2024-02-12 00:00:00 | 0/1/10705   | 2.52849 |   0 | Chronos-Bolt Individual |
|  6 | 2024-02-19 00:00:00 | 0/1/10705   | 2.49153 |   0 | Chronos-Bolt Individual |
|  7 | 2024-02-26 00:00:00 | 0/1/10705   | 2.69589 |   2 | Chronos-Bolt Individual |
|  8 | 2024-03-04 00:00:00 | 0/1/10705   | 3.08985 |   4 | Chronos-Bolt Individual |
|  9 | 2024-03-11 00:00:00 | 0/1/10705   | 3.31218 |   2 | Chronos-Bolt Individual |
```

and 

```
|    | ds                  | unique_id   |    fcst |   y | model              |
|---:|:--------------------|:------------|--------:|----:|:-------------------|
|  0 | 2024-01-08 00:00:00 | 0/1/1149    | 2.85593 |   0 | Chronos-Bolt Batch |
|  1 | 2024-01-15 00:00:00 | 0/1/1149    | 2.75801 |   0 | Chronos-Bolt Batch |
|  2 | 2024-01-22 00:00:00 | 0/1/1149    | 2.65326 |   0 | Chronos-Bolt Batch |
|  3 | 2024-01-29 00:00:00 | 0/1/1149    | 2.53008 |   1 | Chronos-Bolt Batch |
|  4 | 2024-02-05 00:00:00 | 0/1/1149    | 2.43385 |   0 | Chronos-Bolt Batch |
|  5 | 2024-02-12 00:00:00 | 0/1/1149    | 2.52849 |   2 | Chronos-Bolt Batch |
|  6 | 2024-02-19 00:00:00 | 0/1/1149    | 2.49152 |   0 | Chronos-Bolt Batch |
|  7 | 2024-02-26 00:00:00 | 0/1/1149    | 2.69589 |   1 | Chronos-Bolt Batch |
|  8 | 2024-03-04 00:00:00 | 0/1/1149    | 3.08985 |   0 | Chronos-Bolt Batch |
|  9 | 2024-03-11 00:00:00 | 0/1/1149    | 3.31218 |   1 | Chronos-Bolt Batch |
```

**Comment by StatMixedML** [comment]
Yes indeed! Pandas ordered the series alphabetically as compared to the ordering in the context list. 

```python
torch.manual_seed(123)

train_df_sub = train[train["unique_id"].isin(series_ids)]
test_df_sub = test[test["unique_id"].isin(series_ids)]
context_list = [torch.tensor(train[train["unique_id"] == series]["y"].values.reshape(-1,)) for series in series_ids]
q_fcsts_batch, mean_fcsts_batch = chronos_model.predict_quantiles(
    context=context_list,
    prediction_length=fcst_h,
    quantile_levels=quantile_levels,
)

chronos_fcst_batch=[]
for i, series in enumerate(series_ids):
    test_df_sub = test[test["unique_id"] == series]
    chronos_fcst_batch.append(
        pd.DataFrame.from_dict({
            "ds": pd.to_datetime(test_df_sub["ds"].values),
            "unique_id": test_df_sub["unique_id"].values,
            "fcst": q_fcsts_batch[i,:,q_sel].flatten(),
            "y": test_df_sub["y"].values.flatten(),
            "model": f"Chronos-Bolt Batch",
        })
    )

chronos_fcst_batch = pd.concat(chronos_fcst_batch, axis=0)
```python


Running the above now shows that both versions overlap. I`m glad that the error was in my code, not in the bolt-implentation.  

![Image](https://github.com/user-attachments/assets/87b6c008-7b1f-4105-8408-c05344f16791)
